Add Systemd socket activation support

StefanHamminga opened the pull request:
I made a small modification to `app.js` to allow starting compiler explorer by accessing it's socket, using [Systemd socket activation](http://0pointer.de/blog/projects/socket-activation.html). A rather crude time-out system (disabled when set to `0` seconds by environment variable) is set to a default of 5 minutes.
There is a document with templates for the required Systemd socket and service files.

jaredwy commented:
Does this work on windows?

On Sat, 3 Mar 2018 at 8:04 pm codecov[bot] <notifications@github.com> wrote:

> Codecov
> <https://codecov.io/gh/mattgodbolt/compiler-explorer/pull/851?src=pr&el=h1>
> Report
>
> Merging #851
> <https://codecov.io/gh/mattgodbolt/compiler-explorer/pull/851?src=pr&el=desc>
> into master
> <https://codecov.io/gh/mattgodbolt/compiler-explorer/commit/05f9e7c87d9a548b73b551ee04d3726b95777b8e?src=pr&el=desc>
> will *not change* coverage.
> The diff coverage is n/a.
>
> [image: Impacted file tree graph]
> <https://codecov.io/gh/mattgodbolt/compiler-explorer/pull/851?src=pr&el=tree>
>
> @@           Coverage Diff           @@##           master     #851   +/-   ##
> =======================================
>   Coverage   58.47%   58.47%
> =======================================
>   Files          43       43
>   Lines        2066     2066
>   Branches      405      405
> =======================================
>   Hits         1208     1208
>   Misses        858      858
>
> ------------------------------
>
> Continue to review full report at Codecov
> <https://codecov.io/gh/mattgodbolt/compiler-explorer/pull/851?src=pr&el=continue>
> .
>
> *Legend* - Click here to learn more
> <https://docs.codecov.io/docs/codecov-delta>
> Δ = absolute <relative> (impact), ø = not affected, ? = missing data
> Powered by Codecov
> <https://codecov.io/gh/mattgodbolt/compiler-explorer/pull/851?src=pr&el=footer>.
> Last update 05f9e7c...e5e4a04
> <https://codecov.io/gh/mattgodbolt/compiler-explorer/pull/851?src=pr&el=lastupdated>.
> Read the comment docs <https://docs.codecov.io/docs/pull-request-comments>
> .
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mattgodbolt/compiler-explorer/pull/851#issuecomment-370132643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AASIdzexYqDO-YUav6L7EwujZOnVQqitks5talyzgaJpZM4Sa17L>
> .
>

partouf commented:
Seems to work for startup and opening the initial page, looking at the code that should be enough to cover this case.

Would've liked it better if the whitespaces weren't staged in the commit though :$

StefanHamminga commented:
@jaredwy This is Linux only, I don't have a test Windows system right now, but if the library can't obtain a socket from Systemd it will revert to normal operation. I took a quick look at the code of [`systemd-socket`](https://github.com/herzi/systemd-socket) and it seems pretty non-intrusive. If you know of any Windows library to achieve the same I'd be happy to take a look.

@partouf Sorry about the whitespace, the comments must have tripped my automatic clean-up scripts, they were certainly not meant to be part of this.

partouf commented:
Can confirm it keeps working like usual when you don't have it setup via systemd

jaredwy commented on app.js:
```diff
@@ -664,7 +665,26 @@ Promise.all([findCompilers(), aws.initConfig(awsProps)])
+
+        const ss = systemdSocket();
+        if (ss) {
+            const timeout = (typeof process.env.IDLE_TIMEOUT !== 'undefined' ? process.env.IDLE_TIMEOUT : 300) * 1000; // ms (5 min default)
```
Could you turn this into a function please.

mattgodbolt replied:
:+1: to this.

StefanHamminga commented:
@jaredwy and @mattgodbolt, I moved the whole construct into a separate function. I chose to pass the `webServer` to the function rather than returning the socket, as it needs to setup the time-out reset route (and it would be inefficient to have that without Systemd active).

I also moved the "listening on" log statement to the function, imho it's better placed there next to the actual `listen` call and now reports the right port in either case.

partouf commented:
Works on:
- [x] Windows
- [x] Mac

mattgodbolt commented:
Thanks @partouf for checking on our other platforms.
